File: LeetCode/Medium/0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        dummy: Optional[ListNode] = ListNode()
        result: Optional[ListNode] = dummy
        carry: int = 0
        while l1 and l2:
            _ = l1.val + l2.val + carry
            dummy.next = ListNode(_%10)
            carry = _ // 10
            dummy = dummy.next
            l1 = l1.next
            l2 = l2.next
        
        while l1:
            _ = l1.val + carry
            dummy.next = ListNode(_%10)
            carry = _ // 10
            dummy = dummy.next
            l1 = l1.next
        
        while l2:
            _ = l2.val + carry
            dummy.next = ListNode(_%10)
            carry = _ // 10
            dummy = dummy.next
            l2 = l2.next
        
        if carry:
            dummy.next = ListNode(carry)

        return result.next
        

File: LeetCode/Medium/0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {

        ListNode* dummyHead = new ListNode();
        ListNode* dummy = dummyHead;
        int carry = 0;
        int temp;

        while(l1 != nullptr || l2 != nullptr || carry != 0){
            int v1 = (l1 != nullptr) ? l1->val : 0;
            int v2 = (l2 != nullptr) ? l2->val : 0;

            temp = v1 + v2 + carry;
            dummy->next = new ListNode(temp % 10);
            carry = int(temp / 10);

            dummy = dummy->next;
            if(l1 != nullptr){
                l1 = l1->next;
            }
            if(l2 != nullptr){
                l2 = l2->next;
            } 
        }

        return dummyHead->next;
    }
};